src/main/java/org/opensearch/knn/training/TrainingJobClusterStateListener.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ *
+ * Modifications Copyright OpenSearch Contributors. See
+ * GitHub history for details.
+ */
+
+package org.opensearch.knn.training;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.index.IndexResponse;
+import org.opensearch.action.search.SearchRequest;
+import org.opensearch.action.search.SearchResponse;
+import org.opensearch.cluster.ClusterChangedEvent;
+import org.opensearch.cluster.ClusterStateListener;
+import org.opensearch.cluster.node.DiscoveryNode;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.unit.TimeValue;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.knn.indices.Model;
+import org.opensearch.knn.indices.ModelDao;
+import org.opensearch.knn.indices.ModelMetadata;
+import org.opensearch.knn.indices.ModelState;
+import org.opensearch.search.SearchHit;
+import org.opensearch.threadpool.ThreadPool;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.ExecutionException;
+
+public class TrainingJobClusterStateListener implements ClusterStateListener {
+
+    public static Logger logger = LogManager.getLogger(TrainingJobRunner.class);
+    private static TrainingJobClusterStateListener INSTANCE;
+
+    private static ModelDao modelDao;
+    private static ThreadPool threadPool;
+    private static ClusterService clusterService;
+
+    /**
+     * Get singleton instance of TrainingJobRunner
+     *
+     * @return singleton instance of TrainingJobRunner
+     */
+    public static synchronized TrainingJobClusterStateListener getInstance() {
+        if (INSTANCE == null) {
+            INSTANCE = new TrainingJobClusterStateListener();
+        }
+        return INSTANCE;
+    }
+
+    /**
+     * Initializes static components.
+     *
+     * @param threadPool threadPool to use to schedule update of models
+     * @param modelDao modelDao used to get modelIds
+     * @param clusterService clusterService used to add a listener
+     */
+    public static void initialize(ThreadPool threadPool, ModelDao modelDao, ClusterService clusterService) {
+        TrainingJobClusterStateListener.threadPool = threadPool;
+        TrainingJobClusterStateListener.modelDao = modelDao;
+        TrainingJobClusterStateListener.clusterService = clusterService;
+        clusterService.addListener(TrainingJobClusterStateListener.getInstance());
```

Comment:
Can we do this outside of this class? In the plugin class itself after we initialize? This will provide better decoupling for testing.